Adds TQDMProgressBar in callbacks

shun-lin opened the pull request:
Adding TQDMProgressBar in callbacks to make integration of TQDM (for progress bar) with tensorflow easy and configurable.

I also included ipython notebook usage / demo in the docs/tutorial folder.

Original source code:
https://github.com/bstriner/keras-tqdm
https://gist.github.com/ebursztein/5ff21d7600556de9946eac4a313344be

Original Pull Request :
https://github.com/tensorflow/addons/pull/519
(sorry I messed up with merging and thus I have create another repo, the current one, to make pull request).

Based on the comments on 519:
1. I have ran ```bash tools/run_docker.sh -c 'make code-format'``` and all tests (3) passes.
2. I have updated callback README.


Thank you so much for your time!

shun-lin commented:
@seanpmorgan I have fixed a typo that is causing tests to fail, may you add the ```kokoro:force-run``` label again? Thank you so much for your time!

seanpmorgan commented:
> @seanpmorgan I have fixed a typo that is causing tests to fail, may you add the `kokoro:force-run` label again? Thank you so much for your time!

Hmmmm looks like there is still a formatting error that is found. Can you double check that any re-formats were committed/pushed

shun-lin commented:
> > @seanpmorgan I have fixed a typo that is causing tests to fail, may you add the `kokoro:force-run` label again? Thank you so much for your time!
> 
> Hmmmm looks like there is still a formatting error that is found. Can you double check that any re-formats were committed/pushed

@seanpmorgan sorry I forgot to check sanity with ```bash tools/run_docker.sh -c 'make sanity-check'``` I have ran it and updated the code to pass all the sanity check, may you add `kokoro:force-run` label again? Thank you so much for your time!

Another question, I found out that when someone just look at the notebook, the private output of the kernal does not show the progress bar, where should I put the below screenshot (or if I should at all)? If it would be helpful, I am thinking of putting it under `doc/tutorial/assets` and then link it in the notebook. Thank you so much! 
https://drive.google.com/file/d/1EEqoF37tknt5Gtl5GCp9_wDh5CRDqF15/view?usp=sharing

shun-lin commented:
@Squadrick when you have time may you review this PR? Thank you so much for your time!

seanpmorgan commented on tensorflow_addons/callbacks/README.md:
```diff
@@ -3,12 +3,12 @@
+|  tqdm_progress_bar | @shun-lin | shunlin@google.com |
 
 ## Contents
 | Submodule | Metric  | Reference                               |
```
Could you edit this README to say `Callback` instead of `Metric` then you can fill it in with `TQDMProgressBar`

shun-lin replied:
got it :)

seanpmorgan commented on docs/tutorials/tqdm_progress_bar.ipynb:
```diff
@@ -0,0 +1,465 @@
+{
```
<p>Mind removing these warnings from the notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/addons/pull/610/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

shun-lin replied:
Will do, thanks!

seanpmorgan commented on docs/tutorials/tqdm_progress_bar.ipynb:
```diff
@@ -0,0 +1,465 @@
+{
```
<p>Hmmm is this what the expected output looks like? I thought it should show similar to:</p><p><a href="https://i.stack.imgur.com/gpSTt.png" target="_blank">https://i.stack.imgur.com/gpSTt.png</a></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/addons/pull/610/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

shun-lin replied:
Resolved, see the newest commit for the change!

seanpmorgan commented:
Also just want to tag @bstriner. The repo has an MIT license so I think it's fine but wanted to tag for visibility

shun-lin commented:
> Also just want to tag @bstriner. The repo has an MIT license so I think it's fine but wanted to tag for visibility

Thanks Sean!

shun-lin commented:
@seanpmorgan I have updated this Pull Request based on your comments, may you run ```kokoro:force-run``` again and review the PR? Thank you so much for your time!

seanpmorgan commented on docs/tutorials/tqdm_progress_bar.ipynb:
```diff
@@ -0,0 +1,201 @@
+{
```
<p>Could you add <code>!pip install -q --no-deps tensorflow-addons~=0.6</code> we'll use this notebook as the test for the callback so the notebook will need to be able to execute</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/addons/pull/610/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

shun-lin replied:
got it!

shun-lin replied:
updated in the latest patch!

seanpmorgan commented on docs/tutorials/tqdm_progress_bar.ipynb:
```diff
@@ -0,0 +1,201 @@
+{
```
<p>Instead of merging this .png could you just display the URL using markdown?</p><p> </p><p><code>&nbsp;&nbsp;"![fig](<a href="https://raw.githubusercontent.com/tensorflow/addons/59961669a0e21eb4c045d4ad38d008a529d566c2/docs/tutorials/assets/tqdm_progress_bar_demo.png" target="_blank">https://raw.githubusercontent.com/tensorflow/addons/59961669a0e21eb4c045d4ad38d008a529d566c2/docs/tutorials/assets/tqdm_progress_bar_demo.png</a>)"</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/addons/pull/610/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

shun-lin replied:
got it!

shun-lin replied:
updated in latest patch!

seanpmorgan commented on tensorflow_addons/callbacks/tqdm_progress_bar.py:
```diff
@@ -0,0 +1,190 @@
+
+from tensorflow.keras.callbacks import Callback
+from tensorflow_addons.utils import keras_utils
+from tqdm.auto import tqdm
```
Since tqdm isn't a requirement for addons could you put this import as a try-catch within the class that returns an error if tqdm is unable to import

shun-lin replied:
got it, will do!

shun-lin replied:
updated in latest patch!

seanpmorgan commented on docs/tutorials/tqdm_progress_bar.ipynb:
```diff
@@ -0,0 +1,201 @@
+{
```
<p>Can you set this to use TF2:</p><p><code></code><code></p><p>try:</p><p>&nbsp;%tensorflow_version 2.x</p><p>except:</p><p>&nbsp;pass</p><p> </p><p>import tensorflow as tf</p><p></code><code></code></p><p> </p><p>We should be able to remove these in the near future</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/addons/pull/610/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

shun-lin replied:
@seanpmorgan I thought this snippet fetched tensorflow 2.0

Copied from the tensorflow 2.0 for expert codelab
https://colab.sandbox.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/quickstart/advanced.ipynb

shun-lin commented:
> Thanks for the updates. So I tried running this in colab and didn't return the expected result. Any ideas?
> https://colab.research.google.com/drive/1yIRkyVZEbOWtAzr2BKDoyDeCJFJinR7d

I will look into it asap, thanks!

shun-lin commented:
@seanpmorgan I have updated this PR based on your comments and found the solution to the issue mentioned (```tqdm>=4.36.1``` solves it). When you have time may you review it again? Thank you so much for your time!

seanpmorgan commented on tensorflow_addons/callbacks/tqdm_progress_bar.py:
```diff
@@ -0,0 +1,194 @@
+from tensorflow_addons.utils import keras_utils
+
+try:
+    from tqdm.auto import tqdm
```
Also could you move this into the `__init__` of the class. Because all callbacks are imported at import this will cause issues for users that do not intend to user this:
